Mathlib/Tactic/DeriveEncodable.lean

```diff
@@ -0,0 +1,363 @@
+/-
+Copyright (c) 2024 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean.Meta.Transform
+import Lean.Meta.Inductive
+import Lean.Elab.Deriving.Basic
+import Lean.Elab.Deriving.Util
+import Mathlib.Logic.Encodable.Basic
+import Mathlib.Data.Nat.Pairing
+
+/-!
+# `Encodable` deriving handler
+
+Adds a deriving handler for the `Encodable` class.
+-/
+
+namespace Mathlib.Deriving.Encodable
+open Lean Parser.Term Elab Deriving Meta
+
+/-!
+### Theory
+
+The idea is that every encodable inductive type can be represented as a tree of natural numbers.
+Inspiration for this is s-expressions used in Lisp/Scheme.
+```lean
+inductive S : Type where
+  | nat (n : ℕ)
+  | cons (a b : S)
+```
+We start by constructing a equivalence `S ≃ ℕ` using the `Nat.pair` function.
+
+Here is an example of how this module constructs an encoding.
+
+Suppose we are given the following type:
+```lean
+inductive T (α : Type) where
+  | a (x : α) (y : Bool) (z : T α)
+  | b
+```
+The deriving handler constructs the following declarations:
+```lean
+def encodableT_toS {α} [Encodable α] (x : T α) : S :=
+  match x with
+  | T.a a a_1 a_2 =>
+    S.cons (S.nat 0)
+      (S.cons (S.nat (Encodable.encode a))
+        (S.cons (S.nat (Encodable.encode a_1)) (S.cons (encodableT_toS a_2) (S.nat 0))))
+  | T.b => S.cons (S.nat 1) (S.nat 0)
+
+private def encodableT_fromS {α} [Encodable α] : S → Option (T α) := fun
+  | S.cons (S.nat 0) (S.cons (S.nat a) (S.cons (S.nat a_1) (S.cons a_2 (S.nat 0)))) =>
+    match Encodable.decode a, Encodable.decode a_1, encodableT_fromS a_2 with
+    | some a, some a_1, some a_2 => some <| T.a a a_1 a_2
+    | _, _, _ => none
+  | S.cons (S.nat 1) (S.nat 0) => some <| T.b
+  | _ => none
+
+private theorem encodableT {α} [Encodable α] (x : @T α) :
+    encodableT_fromS (encodableT_toS x) = some x := by
+  cases x <;> (unfold encodableT_toS encodableT_fromS; simp only [Encodable.encodek, encodableT])
+
+instance {α} [Encodable α] : Encodable (@T α) :=
+  Encodable.ofLeftInjection encodableT_toS encodableT_fromS encodableT
+```
+The idea is that each constructor gets encoded as a linked list made of `S.cons` constructors
+that is tagged with the constructor index.
+-/
+
+private inductive S : Type where
```

Comment:
Given that `Encodable` contains data that mentions `S`, we might conceivably want to write lemmas about `S` and `S.encode`. Maybe give a proper name and make it non-private? (or not give it a proper name and make it `protected`)

Reply:
Let's consider the actual encoding to be an implementation detail, since it's subject to change.

I've added a sentence to the module doc about it being an implementation detail.